cluster-autoscaler/main.go

```diff
@@ -250,38 +249,39 @@ func registerSignalHandlers(autoscaler core.Autoscaler) {
 	}()
 }
 
-func run(healthCheck *metrics.HealthCheck) {
-	metrics.RegisterAll()
-	kubeConfig := getKubeConfig()
-	kubeClient := createKubeClient(kubeConfig)
-	kubeEventRecorder := kube_util.CreateEventRecorder(kubeClient)
+func buildAutoscaler() (core.Autoscaler, error) {
+	// Create basic config from flags.
 	autoscalingOptions := createAutoscalingOptions()
-	metrics.UpdateNapEnabled(autoscalingOptions.NodeAutoprovisioningEnabled)
-	predicateCheckerStopChannel := make(chan struct{})
-	predicateChecker, err := simulator.NewPredicateChecker(kubeClient, predicateCheckerStopChannel)
-	if err != nil {
-		glog.Fatalf("Failed to create predicate checker: %v", err)
-	}
-	listerRegistryStopChannel := make(chan struct{})
-	listerRegistry := kube_util.NewListerRegistryWithDefaultListers(kubeClient, listerRegistryStopChannel)
-
+	kubeClient := createKubeClient(getKubeConfig())
 	opts := core.AutoscalerOptions{
 		AutoscalingOptions: autoscalingOptions,
-		PredicateChecker:   predicateChecker,
 		KubeClient:         kubeClient,
-		KubeEventRecorder:  kubeEventRecorder,
-		ListerRegistry:     listerRegistry,
 	}
 
-	cloudProvider := cloudBuilder.NewCloudProvider(autoscalingOptions)
+	// This metric should be published only once.
+	metrics.UpdateNapEnabled(autoscalingOptions.NodeAutoprovisioningEnabled)
+
+	// Create autoscaler.
+	return core.NewAutoscaler(opts)
+}
+
+func run(healthCheck *metrics.HealthCheck) {
+	// Register metrics.
+	metrics.RegisterAll()
 
-	autoscaler, err := core.NewAutoscaler(opts, cloudProvider)
+	// Create autoscaler.
+	autoscaler, err := buildAutoscaler()
 	if err != nil {
 		glog.Fatalf("Failed to create autoscaler: %v", err)
 	}
+
+	// Register signal handlers for graceful shutdown.
```

Comment:
This comment I like, because there is an additional information on why we need signal handlers.